File: 5-longest-palindromic-substring/5-longest-palindromic-substring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        output = ""
        for i in range(2 * len(s) - 1): # O(n)
            left, right = i // 2, (i + 1) // 2
            while left >= 0 and right < len(s): # O(n)
                if s[left] != s[right]:
                    break
                if right - left + 1 > len(output):
                    output = s[left:right + 1]
                left -= 1
                right += 1
            if 2 * len(s) - i - 1 < len(output):
                break
        return output

File: 5-longest-palindromic-substring/5-longest-palindromic-substring.java
class Solution {
    public String longestPalindrome(String s) {
        int start = 0, end = 0;
        char[] ca = s.toCharArray(); // apparently array access >> string random access
        for (int i = 0; i < ca.length; ++i) {
            char c = ca[i];
            int left = i;
            int right = i;
            while (left >= 0 && ca[left] == c) --left;
            while (right < ca.length && ca[right] == c) ++right;
            while (left >= 0 && right < ca.length) {
                if (ca[left] != ca[right]) break;
                --left;
                ++right;
            }
            ++left;
            if (end - start < right - left) {
                start = left;
                end = right;
            }
        }
        return s.substring(start, end);
    }
}